File: src/main/java/com/cloudpid/ai/agents/AnthropicProvider.java
package com.cloudpid.ai.agents;

import com.anthropic.client.AnthropicClient;
import com.anthropic.client.okhttp.AnthropicOkHttpClient;
import com.anthropic.models.messages.*;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.*;
import java.util.stream.Collectors;

/** Anthropic API provider — Anthropic API + server-side web_search_20260209. */
final class AnthropicProvider implements Provider {

    private static final Logger log = LoggerFactory.getLogger(AnthropicProvider.class);
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final long MAX_TOKENS = 16000L;

    private final AnthropicClient client;
    private final String model;
    private final String system;
    private final List<Tool> customTools;
    private final List<MessageParam> messages = new ArrayList<>();

    AnthropicProvider(AnthropicClient client, String model, String system,
                      String kickoff, List<Tool> customTools) {
        this.client = client;
        this.model = model;
        this.system = system;
        this.customTools = List.copyOf(customTools);
        messages.add(MessageParam.builder()
            .role(MessageParam.Role.USER)
            .content(kickoff)
            .build());
    }

    static Provider create(String model, String system, String kickoff, List<Tool> customTools) {
        return new AnthropicProvider(AnthropicOkHttpClient.fromEnv(), model, system, kickoff, customTools);
    }

    @Override
    public AgentResponse nextTurn() {
        MessageCreateParams.Builder builder = MessageCreateParams.builder()
            .model(model)
            .maxTokens(MAX_TOKENS)
            .thinking(ThinkingConfigAdaptive.builder().build())
            .system(system)
            .addTool(WebSearchTool20260209.builder().build())
            .messages(messages);
        customTools.forEach(builder::addTool);

        Message response = client.messages().create(builder.build());
        log.debug("stop_reason={}", response.stopReason());

        messages.add(toAssistantParam(response));

        List<ToolCall> toolCalls = response.content().stream()
            .filter(b -> b.toolUse().isPresent())
            .map(b -> {
                ToolUseBlock tu = b.toolUse().get();
                return new ToolCall(tu.id(), tu.name(), parseInput(tu));
            })
            .collect(Collectors.toList());

        return new AgentResponse(stopReasonString(response.stopReason()), toolCalls);
    }

    @Override
    public void submitToolResults(List<Map.Entry<ToolCall, String>> results) {
        List<ContentBlockParam> blocks = results.stream()
            .map(e -> ContentBlockParam.ofToolResult(
                ToolResultBlockParam.builder()
                    .toolUseId(e.getKey().id())
                    .content(e.getValue())
                    .build()))
            .collect(Collectors.toList());
        messages.add(MessageParam.builder()
            .role(MessageParam.Role.USER)
            .contentOfBlockParams(blocks)
            .build());
    }

    private static MessageParam toAssistantParam(Message response) {
        List<ContentBlockParam> blocks = new ArrayList<>();
        for (ContentBlock block : response.content()) {
            block.thinking().ifPresent(t ->
                blocks.add(ContentBlockParam.ofThinking(
                    ThinkingBlockParam.builder()
                        .thinking(t.thinking())
                        .signature(t.signature())
                        .build())));
            block.text().ifPresent(t ->
                blocks.add(ContentBlockParam.ofText(
                    TextBlockParam.builder().text(t.text()).build())));
            block.toolUse().ifPresent(tu ->
                blocks.add(ContentBlockParam.ofToolUse(
                    ToolUseBlockParam.builder()
                        .id(tu.id())
                        .name(tu.name())
                        .input(tu._input())
                        .build())));
        }
        return MessageParam.builder()
            .role(MessageParam.Role.ASSISTANT)
            .contentOfBlockParams(blocks)
            .build();
    }

    private static Map<String, Object> parseInput(ToolUseBlock tu) {
        try {
            return MAPPER.readValue(tu._input().toString(),
                new TypeReference<Map<String, Object>>() {});
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private static String stopReasonString(Optional<StopReason> opt) {
        StopReason sr = opt.orElse(null);
        if (StopReason.TOOL_USE.equals(sr)) return "tool_use";
        if (StopReason.PAUSE_TURN.equals(sr)) return "pause_turn";
        if (StopReason.MAX_TOKENS.equals(sr)) return "max_tokens";
        return "end_turn";
    }
}

File: src/main/rust/src/agents/anthropic_provider.rs
use crate::agents::provider::{AgentResponse, Provider, ToolCall};
use anyhow::{anyhow, Result};
use async_trait::async_trait;
use reqwest::Client;
use serde_json::{json, Value};
use std::collections::HashMap;
use tracing::debug;

const API_URL: &str = "https://api.anthropic.com/v1/messages";
const DEFAULT_MODEL: &str = "claude-sonnet-4-6";
const MAX_TOKENS: u64 = 16000;
const API_VERSION: &str = "2023-06-01";

pub struct AnthropicProvider {
    client: Client,
    api_key: String,
    model: String,
    system: String,
    tools: Vec<Value>,
    messages: Vec<Value>,
}

impl AnthropicProvider {
    pub fn new(
        api_key: String,
        model: impl Into<String>,
        system: impl Into<String>,
        kickoff: impl Into<String>,
        custom_tools: Vec<Value>,
    ) -> Self {
        let mut tools = vec![json!({"type": "web_search_20260209", "name": "web_search"})];
        tools.extend(custom_tools);

        Self {
            client: Client::new(),
            api_key,
            model: model.into(),
            system: system.into(),
            tools,
            messages: vec![json!({"role": "user", "content": kickoff.into()})],
        }
    }

    pub fn from_env(
        system: impl Into<String>,
        kickoff: impl Into<String>,
        custom_tools: Vec<Value>,
    ) -> Result<Self> {
        let api_key = std::env::var("ANTHROPIC_API_KEY")
            .map_err(|_| anyhow!("ANTHROPIC_API_KEY environment variable not set"))?;
        Ok(Self::new(api_key, DEFAULT_MODEL, system, kickoff, custom_tools))
    }
}

#[async_trait]
impl Provider for AnthropicProvider {
    async fn next_turn(&mut self) -> Result<AgentResponse> {
        let body = json!({
            "model": self.model,
            "max_tokens": MAX_TOKENS,
            "thinking": {"type": "adaptive"},
            "system": self.system,
            "tools": self.tools,
            "messages": self.messages,
        });

        let response = self
            .client
            .post(API_URL)
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", API_VERSION)
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            return Err(anyhow!("Anthropic API error {}: {}", status, text));
        }

        let resp: Value = response.json().await?;
        let stop_reason = resp["stop_reason"]
            .as_str()
            .unwrap_or("end_turn")
            .to_string();
        debug!(stop_reason = %stop_reason);

        // Append assistant turn to history (full content including thinking blocks).
        self.messages
            .push(json!({"role": "assistant", "content": resp["content"].clone()}));

        let tool_calls = resp["content"]
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter(|b| b["type"].as_str() == Some("tool_use"))
                    .map(|b| ToolCall {
                        id: b["id"].as_str().unwrap_or("").to_string(),
                        name: b["name"].as_str().unwrap_or("").to_string(),
                        input: b["input"]
                            .as_object()
                            .map(|m| {
                                m.iter()
                                    .map(|(k, v)| (k.clone(), v.clone()))
                                    .collect::<HashMap<_, _>>()
                            })
                            .unwrap_or_default(),
                    })
                    .collect()
            })
            .unwrap_or_default();

        Ok(AgentResponse {
            stop_reason,
            tool_calls,
        })
    }

    async fn submit_tool_results(&mut self, results: Vec<(ToolCall, String)>) -> Result<()> {
        let content: Vec<Value> = results
            .into_iter()
            .map(|(tc, result)| {
                json!({
                    "type": "tool_result",
                    "tool_use_id": tc.id,
                    "content": result,
                })
            })
            .collect();
        self.messages
            .push(json!({"role": "user", "content": content}));
        Ok(())
    }
}
